node-repository/src/main/java/com/yahoo/vespa/hosted/provision/autoscale/ClusterTimeseries.java

```diff
@@ -1,70 +1,103 @@
 
 /**
- * A series of metric snapshots for all nodes in a cluster
+ * A list of metric snapshots from a cluster, sorted by increasing time (newest last).
  *
  * @author bratseth
  */
 public class ClusterTimeseries {
 
-    private final NodeList clusterNodes;
+    private final ClusterSpec.Id cluster;
+    private final List<ClusterMetricSnapshot> snapshots;
 
-    /** The measurements for all nodes in this snapshot */
-    private final List<NodeTimeseries> allTimeseries;
+    ClusterTimeseries(ClusterSpec.Id cluster, List<ClusterMetricSnapshot> snapshots) {
+        this.cluster = cluster;
+        List<ClusterMetricSnapshot> sortedSnapshots = new ArrayList<>(snapshots);
+        Collections.sort(sortedSnapshots);
+        this.snapshots = Collections.unmodifiableList(sortedSnapshots);
+    }
+
+    public boolean isEmpty() { return snapshots.isEmpty(); }
+
+    public int size() { return snapshots.size(); }
+
+    public ClusterMetricSnapshot get(int index) { return snapshots.get(index); }
 
-    public ClusterTimeseries(Duration period, Cluster cluster, NodeList clusterNodes, MetricsDb db) {
-        this.clusterNodes = clusterNodes;
-        var timeseries = db.getNodeTimeseries(period, clusterNodes);
+    public List<ClusterMetricSnapshot> asList() { return snapshots; }
 
-        if (cluster.lastScalingEvent().isPresent())
-            timeseries = filter(timeseries, snapshot -> snapshot.generation() < 0 || // Content nodes do not yet send generation
-                                                        snapshot.generation() >= cluster.lastScalingEvent().get().generation());
-        timeseries = filter(timeseries, snapshot -> snapshot.inService() && snapshot.stable());
+    public ClusterSpec.Id cluster() { return cluster; }
 
-        this.allTimeseries = timeseries;
+    public ClusterTimeseries add(ClusterMetricSnapshot snapshot) {
+        List<ClusterMetricSnapshot> list = new ArrayList<>(snapshots);
+        list.add(snapshot);
+        return new ClusterTimeseries(cluster, list);
     }
 
-    /** Returns the average number of measurements per node */
-    public int measurementsPerNode() {
-        int measurementCount = allTimeseries.stream().mapToInt(m -> m.size()).sum();
-        return measurementCount / clusterNodes.size();
+    /** The max query growth rate we can predict from this time-series as a fraction of the current traffic per minute */
+    public double maxQueryGrowthRate() {
+        if (snapshots.isEmpty()) return 0.1;
+
+        // Find the period having the highest growth rate, where total growth exceeds 30% increase
+        double maxGrowthRate = 0; // In query rate per minute
+        for (int start = 0; start < snapshots.size(); start++) {
+            if (start > 0) { // Optimization: Skip this point when starting from the previous is better relative to the best rate so far
+                Duration duration = durationBetween(start - 1, start);
+                if ( ! duration.isZero()) {
```

Comment:
When is `duration.isZero()`? Shouldn't we check if `duration.toMinutes() == 0` instead? Same on line 67

Reply:
Good point; done. (This could happen if we end up with duplicate reads of the same metric window and nothing else.)